src/io/flutter/bazel/WorkspaceCache.java

```diff
@@ -64,46 +82,82 @@ public static WorkspaceCache getInstance(@NotNull final Project project) {
    * Returns the Workspace in the cache.
    * <p>
    * <p>Returning a null means there is no current workspace for this project.
-   * <p>
-   * <p>If the cache hasn't loaded yet, blocks until it's ready.
-   * Otherwise doesn't block.
    */
   @Nullable
-  public Workspace getNow() {
-    return cache.getNow();
+  public Workspace get() {
+    return cache;
   }
 
   /**
-   * Waits for any refreshes to finish, then returns the new workspace (or null).
-   *
-   * @throws IllegalStateException if called on the Swing dispatch thread.
+   * Returns whether the  project is a bazel project.
    */
-  @Nullable
-  public Workspace getWhenReady() {
-    return cache.getWhenReady();
+  public boolean isBazel() {
+    return cache != null;
   }
 
   /**
    * Runs a callback each time the current Workspace changes.
    */
   public void subscribe(Runnable callback) {
-    cache.subscribe(callback);
+    synchronized (subscribers) {
+      subscribers.add(callback);
+    }
   }
 
   /**
    * Stops notifications to a callback passed to {@link #subscribe}.
    */
   public void unsubscribe(Runnable callback) {
-    cache.unsubscribe(callback);
+    synchronized (subscribers) {
+      subscribers.remove(callback);
+    }
   }
 
   /**
-   * Triggers a cache refresh.
+   * The Dart plugin uses this registry key to avoid bazel users getting their settings overridden on projects that include a
+   * pubspec.yaml.
    * <p>
-   * If a refresh is already in progress, schedules another one.
+   * In other words, this key tells the plugin to configure dart projects without pubspec.yaml.
+   */
+  private static final String dartProjectsWithoutPubspecRegistryKey = "dart.projects.without.pubspec";
+
+  /**
+   * Executes a cache refresh.
    */
-  private void refreshAsync() {
-    cache.refresh(() -> Workspace.load(project));
+  private void refresh() {
+    final Workspace workspace = Workspace.loadUncached(project);
+    if (workspace == cache) return;
+    cache = workspace;
+
+    // If the current workspace is a bazel workspace, update the Dart plugin
+    // registry key to indicate that there are dart projects without pubspec
+    // registry keys. TODO(jacobr): it would be nice if the Dart plugin was
+    // instead smarter about handling Bazel projects.
+    if (cache != null) {
+      if (!Registry.is(dartProjectsWithoutPubspecRegistryKey, false)) {
```

Comment:
Does this indicate to the Dart plugin that all projects are bazel projects? Or that the current one might be?

We should probably push some version of bazel `Workspace` knowledge down to the Dart plugin; it'll then have unify behavior with the Flutter plugin and the analysis server. cc @jwren

Reply:
Changing this registry changes it for all projects. This registry disables features such as notifications around pubspec files themselves and disabling notifications around new available Dart SDK versions.

From my perspective having Flutter set this registry, is unifying the behavior between the Dart and Flutter plugin.  We considered the idea of moving a bunch of the `Workspace` data down to the Dart plugin but  Jacob made the point that the in the Flutter plugin, `Workspace` manages Flutter specific bazel data, mainly changes in executing flutter applications, i.e. the json file knowledge

Reply:
I wasn't suggesting moving the specific `Workspace` or `WorkspaceCache` class down to the Dart plugin, but instead removing the flag, and having a similar class in the dart plugin which detects when its in a bazel workspace, and configures the UI to operate slightly differently in those situations.